quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java

```diff
@@ -107,6 +111,77 @@ private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel docu
 		}
 	}
 
+	/**
+	 * Returns the <code>Position</code> to insert the missing required code action property into
+	 * @param textDocument the text document
+	 * @return the <code>Position</code> to insert the missing required code action property into
+	 * @throws BadLocationException
+	 */
+	private Position getPositionForRequiredCodeAction(TextDocument textDocument) throws BadLocationException {
+		String textDocumentText = textDocument.getText();
+
+		Position position;
+		
+		if (!StringUtils.hasText(textDocumentText)) {
+			position = new Position(0, 0);
+		} else {
+			position = textDocument.positionAt(StringUtils.trimEnd(textDocumentText).length());
+		}
+		return position;
+	}
+
+	/**
+	 * Returns the missing required property name from <code>diagnosticMessage</code>
+	 * @param diagnosticMessage the diagnostic message containing the property name in single quotes
+	 * @return the missing required property name from <code>diagnosticMessage</code>
+	 */
+	private String getPropertyNameFromRequiredMessage(String diagnosticMessage) {
+		String property = diagnosticMessage.substring(diagnosticMessage.indexOf("'") + 1);
+		return property.substring(0, property.indexOf("'"));
+	}
+
+	/**
+	 * Create a code action that inserts all missing required properties and equals signs if there
+	 * are more than one missing required properties.
+	 * 
+	 * @param diagnostics        the diagnostics, one for each missing required property
+	 * @param document           the properties model
+	 * @param formattingSettings the formatting settings
+	 * @param codeActions        the code actions list to fill
+	 */
+	private void doCodeActionForAllRequired(List<Diagnostic> diagnostics, PropertiesModel document,
+			QuarkusFormattingSettings formattingSettings, List<CodeAction> codeActions) {
+		
+		TextDocument textDocument = document.getDocument();
+		List<String> properties = new ArrayList<>();

```

Comment:
You are looping twice diagnostics (for computing properties list and for create code action with the given  diagnostic list).  

Instead of computing list of properties, I will collect diagnostics which have required diagnostic 

`List<Diagnostic> requiredDiagnostics = diagnostics.stream().filter(d -> ValidationType.required.name().equals(d.getCode())).collect(Collectors.toList()))`